includes/polling/os/ios.inc.php

```diff
@@ -15,9 +15,12 @@
 } elseif (preg_match('/Cisco Internetwork Operating System Software\s+IOS \(tm\) [^ ]+ Software \([^\-]+-([^\-]+)-\w\),.+?Version ([^, ]+)/', $device['sysDescr'], $regexp_result)) {
     $features = $regexp_result[1];
     $version  = $regexp_result[2];
+} elseif (preg_match('/^Cisco IOS Software \[([^\]]+)\],.+Software \(([^\)]+)\), Version ([^, ]+)/', $device['sysDescr'], $regexp_result)) {
+    $features = $regexp_result[1];
+    $version  = $regexp_result[2] . " " . $regexp_result[3];
```

Comment:
This block breaks compatibility and consistency between different devices.

example on 3 devices:
device 1, catalist 2950 (IOS), sysDescr:
```
Cisco Internetwork Operating System Software\nIOS (tm) C2950 Software (C2950-I6K2L2Q4-M), Version 12.1(22)EA13, RELEASE SOFTWARE (fc2)\r\nTechnical Support: http:\/\/www.cisco.com\/techsupport\r\nCopyright (c) 1986-2009 by cisco Systems, Inc.\r\nCompiled Fri 27-F
```

device 2, catalist 2960 (IOS), sysDescr:
```
Cisco IOS Software, C2960 Software (C2960-LANBASEK9-M), Version 15.0(2)SE8, RELEASE SOFTWARE (fc1)\nTechnical Support: http:\/\/www.cisco.com\/techsupport\r\nCopyright (c) 1986-2015 by Cisco Systems, Inc.\r\nCompiled Thu 14-May-15 02:39 by prod_rel_team
```

device 3, catalist 3850 (IOSXE), sysDescr:
```
Cisco IOS Software [Denali], Catalyst L3 Switch Software (CAT3K_CAA-UNIVERSALK9-M), Version 16.3.8, RELEASE SOFTWARE (fc3)\nTechnical Support: http:\/\/www.cisco.com\/techsupport\r\nCopyright (c) 1986-2019 by Cisco Systems, Inc.\r\nCompiled Wed 13-Feb-19 03:00 b
```

On Cisco IOS devices (and before the change, IOSXE), the 'features' field looks like 'LANBASEK9', 'SPSERVICESK9'. This is a set of functions implemented in the software image. In IOSXE, before this change, it had the same look, for example: 'UNIVERSALK9'. After the change, the 'features' field has become 'Denali'. But this is not quite a feature, it is a human-readable name for a software branch, but it does not give much to understand what functions are implemented in the software. And 'features' is 'UNIVERSALK9'.

The 'version' field has also changed. For IOS devices, it has the form '12 .1 (22) EA13 ', '15 .0 (2) SE8'. For IOSXE it used to be '16 .3.8 ', now it is' CAT3K_CAA-UNIVERSALK9-M 16.3.8'. 'CAT3K_CAA' is actually a platform, it is unchanged for the hardware platform and family, and is contained in the 'hardware' field. 'UNIVERSALK9' is a feature set, it should be in the 'features' field.

Thus, not only compatibility is broken, but the semantic meaning of the fields has changed, which is now different for different types of devices.

field values for different devices (their sysDescr above):

| device | feature | version |
--- | --- | ---
| device 1, catalist 2950 (IOS) | I6K2L2Q4 | 12.1(22)EA13 |
| device 2, catalist 2960 (IOS) | LANBASEK9 | 15.0(2)SE8 |
| device 3, catalist 3850 (IOSXE), before change | UNIVERSALK9 | 16.3.8 |
| device 3, catalist 3850 (IOSXE), after change | Denali | CAT3K_CAA-UNIVERSALK9-M 16.3.8 |

To fix it, just drop the [line](https://github.com/librenms/librenms/pull/10862/files#diff-9444c744797a230750e46a0fb83cbc8aR18). [Line](https://github.com/librenms/librenms/pull/10862/files#diff-9444c744797a230750e46a0fb83cbc8aR12) provided the old behavior for IOSXE.

If someone believes that 'CAT3K_CAA' is an important part of version information, then this part is more logical to combine with 'features', since in the real world the set of implemented functions depends on the hardware platform and knowledge of the platform is a refinement for the implemented feature set. Thus, 'feature' can be of the form 'CAT3K_CAA-UNIVERSALK9'.

Reply:
Hi @seros1521 
Will have a look at this. We don't have much test data so this kind of regression cannot be detected. And cisco does it different on every switch family (and sometimes inside a single switch family) making it very difficult. 
Bye

Reply:
There is still a file [cisco.inc.php](https://github.com/librenms/librenms/blob/78d017d1ba6076c0a5b32aa23dfa217878f58b4f/includes/polling/os/cisco.inc.php). It has almost the same logic as in ios.inc.php and they can overlap each other.

Reply:
cisco (group) is only called if ios (os) or iosxe (os) do not exist. In current state, cisco.inc.php is not used for IOS and IOSXE at all.